File: backend/speed_reading/scoring.py
from __future__ import annotations

import re
from difflib import SequenceMatcher

from .models import Passage, ScoreResult


WORD_RE = re.compile(r"[a-zA-Z']+")


def normalize_words(text: str) -> list[str]:
    return [word.lower().strip("'") for word in WORD_RE.findall(text)]


def score_comprehension(passage: Passage, response: str, pass_threshold: int = 70) -> ScoreResult:
    response_words = normalize_words(response)
    passage_words = normalize_words(passage.content)
    response_text = " ".join(response_words)

    length_points = _score_length(
        len(response_words), passage.comprehension.minimum_response_words
    )
    keyword_points = _score_keywords(
        response_text,
        passage.comprehension.required_keywords,
        passage.comprehension.synonyms,
    )
    concept_points = _score_concepts(response_text, passage.comprehension.concepts)
    originality_points = _score_originality(
        response_words, passage_words, passage.comprehension.copy_limit
    )
    coherence_points = _score_coherence(response_words)

    total = (
        length_points
        + keyword_points
        + concept_points
        + originality_points
        + coherence_points
    )
    feedback = _feedback(
        total,
        length_points,
        keyword_points,
        concept_points,
        originality_points,
        coherence_points,
    )
    return ScoreResult(
        score=total,
        passed=total >= pass_threshold,
        length_points=length_points,
        keyword_points=keyword_points,
        concept_points=concept_points,
        originality_points=originality_points,
        coherence_points=coherence_points,
        feedback=feedback,
    )


def _score_length(word_count: int, minimum: int) -> int:
    if word_count >= minimum:
        return 15
    return round(15 * (word_count / minimum)) if minimum else 0


def _score_keywords(
    response_text: str, required_keywords: list[str], synonyms: dict[str, list[str]]
) -> int:
    if not required_keywords:
        return 0
    matches = 0
    for keyword in required_keywords:
        accepted_terms = [keyword.lower(), *[term.lower() for term in synonyms.get(keyword, [])]]
        if any(_contains_term(response_text, term) for term in accepted_terms):
            matches += 1
    return round(30 * (matches / len(required_keywords)))


def _score_concepts(response_text: str, concepts: list[str]) -> int:
    if not concepts:
        return 0
    matched = 0
    for concept in concepts:
        concept_words = set(normalize_words(concept))
        if not concept_words:
            continue
        overlap = sum(1 for word in concept_words if _contains_term(response_text, word))
        if overlap / len(concept_words) >= 0.45:
            matched += 1
    return round(35 * (matched / len(concepts)))


def _score_originality(
    response_words: list[str], passage_words: list[str], copy_limit: float
) -> int:
    if not response_words:
        return 0
    response_text = " ".join(response_words)
    passage_text = " ".join(passage_words)
    similarity = SequenceMatcher(None, response_text, passage_text).ratio()
    if similarity <= copy_limit:
        return 10
    if similarity >= 0.9:
        return 0
    return round(10 * ((0.9 - similarity) / (0.9 - copy_limit)))


def _score_coherence(response_words: list[str]) -> int:
    unique_ratio = len(set(response_words)) / len(response_words) if response_words else 0
    if len(response_words) < 5:
        return 0
    if unique_ratio < 0.35:
        return 3
    return 10


def _contains_term(text: str, term: str) -> bool:
    escaped = re.escape(term.lower())
    return re.search(rf"\b{escaped}\b", text) is not None


def _feedback(
    total: int,
    length_points: int,
    keyword_points: int,
    concept_points: int,
    originality_points: int,
    coherence_points: int,
) -> list[str]:
    notes: list[str] = []
    if total >= 70:
        notes.append("Good recall. You captured enough of the passage to move ahead.")
    else:
        notes.append("Try again and include the main idea plus two important details.")
    if length_points < 15:
        notes.append("Write a little more so your answer has enough detail.")
    if keyword_points < 20:
        notes.append("Mention more important names, objects, or actions from the passage.")
    if concept_points < 24:
        notes.append("Focus on what happened and what the passage teaches.")
    if originality_points < 7:
        notes.append("Use your own words instead of copying the passage.")
    if coherence_points < 10:
        notes.append("Write complete, clear sentences.")
    return notes

File: lib/scoring.ts
// TypeScript port of backend/speed_reading/scoring.py.
// Keep the point weights and thresholds in sync with the Python module.

export type ComprehensionConfig = {
  minimumResponseWords: number;
  requiredKeywords: string[];
  concepts: string[];
  synonyms: Record<string, string[]>;
  copyLimit: number;
};

export type PassageData = {
  // SR-R1-003: approved-file-only content - every passage must declare these to be loadable.
  content_id: string;
  content_version: string;
  schema_version: string;
  approval_status: "WIP" | "APPROVED" | "RETIRED";
  level: number;
  title: string;
  category: string;
  difficulty: string;
  estimatedAgeRange: string;
  wordCount: number;
  content: string;
  comprehension: ComprehensionConfig;
};

export type ScoreResult = {
  score: number;
  passed: boolean;
  lengthPoints: number;
  keywordPoints: number;
  conceptPoints: number;
  originalityPoints: number;
  coherencePoints: number;
  feedback: string[];
};

const WORD_RE = /[a-zA-Z']+/g;

export function normalizeWords(text: string): string[] {
  return (text.match(WORD_RE) ?? []).map((word) => word.toLowerCase().replace(/^'+|'+$/g, ""));
}

export function scoreComprehension(
  passage: PassageData,
  response: string,
  passThreshold = 70
): ScoreResult {
  const responseWords = normalizeWords(response);
  const passageWords = normalizeWords(passage.content);
  const responseText = responseWords.join(" ");

  const lengthPoints = scoreLength(responseWords.length, passage.comprehension.minimumResponseWords);
  const keywordPoints = scoreKeywords(
    responseText,
    passage.comprehension.requiredKeywords,
    passage.comprehension.synonyms
  );
  const conceptPoints = scoreConcepts(responseText, passage.comprehension.concepts);
  const originalityPoints = scoreOriginality(
    responseWords,
    passageWords,
    passage.comprehension.copyLimit
  );
  const coherencePoints = scoreCoherence(responseWords);

  const total = lengthPoints + keywordPoints + conceptPoints + originalityPoints + coherencePoints;
  return {
    score: total,
    passed: total >= passThreshold,
    lengthPoints,
    keywordPoints,
    conceptPoints,
    originalityPoints,
    coherencePoints,
    feedback: buildFeedback(
      total,
      lengthPoints,
      keywordPoints,
      conceptPoints,
      originalityPoints,
      coherencePoints
    )
  };
}

function scoreLength(wordCount: number, minimum: number): number {
  if (wordCount >= minimum) return 15;
  return minimum ? Math.round(15 * (wordCount / minimum)) : 0;
}

function scoreKeywords(
  responseText: string,
  requiredKeywords: string[],
  synonyms: Record<string, string[]>
): number {
  if (!requiredKeywords.length) return 0;
  let matches = 0;
  for (const keyword of requiredKeywords) {
    const acceptedTerms = [keyword.toLowerCase(), ...(synonyms[keyword] ?? []).map((t) => t.toLowerCase())];
    if (acceptedTerms.some((term) => containsTerm(responseText, term))) {
      matches += 1;
    }
  }
  return Math.round(30 * (matches / requiredKeywords.length));
}

function scoreConcepts(responseText: string, concepts: string[]): number {
  if (!concepts.length) return 0;
  let matched = 0;
  for (const concept of concepts) {
    const conceptWords = new Set(normalizeWords(concept));
    if (!conceptWords.size) continue;
    let overlap = 0;
    for (const word of conceptWords) {
      if (containsTerm(responseText, word)) overlap += 1;
    }
    if (overlap / conceptWords.size >= 0.45) matched += 1;
  }
  return Math.round(35 * (matched / concepts.length));
}

function scoreOriginality(
  responseWords: string[],
  passageWords: string[],
  copyLimit: number
): number {
  if (!responseWords.length) return 0;
  const similarity = sequenceSimilarity(responseWords.join(" "), passageWords.join(" "));
  if (similarity <= copyLimit) return 10;
  if (similarity >= 0.9) return 0;
  return Math.round(10 * ((0.9 - similarity) / (0.9 - copyLimit)));
}

function scoreCoherence(responseWords: string[]): number {
  if (responseWords.length < 5) return 0;
  const uniqueRatio = new Set(responseWords).size / responseWords.length;
  if (uniqueRatio < 0.35) return 3;
  return 10;
}

function containsTerm(text: string, term: string): boolean {
  const escaped = term.toLowerCase().replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(`\\b${escaped}\\b`).test(text);
}

// Ratcliff-Obershelp similarity, matching Python's difflib.SequenceMatcher.ratio().
function sequenceSimilarity(a: string, b: string): number {
  if (!a.length && !b.length) return 1;
  return (2 * matchingCharacters(a, b, 0, a.length, 0, b.length)) / (a.length + b.length);
}

function matchingCharacters(
  a: string,
  b: string,
  aLow: number,
  aHigh: number,
  bLow: number,
  bHigh: number
): number {
  let bestI = aLow;
  let bestJ = bLow;
  let bestSize = 0;
  let runLengths = new Map<number, number>();
  for (let i = aLow; i < aHigh; i++) {
    const newRunLengths = new Map<number, number>();
    for (let j = bLow; j < bHigh; j++) {
      if (a[i] !== b[j]) continue;
      const k = (runLengths.get(j - 1) ?? 0) + 1;
      newRunLengths.set(j, k);
      if (k > bestSize) {
        bestI = i - k + 1;
        bestJ = j - k + 1;
        bestSize = k;
      }
    }
    runLengths = newRunLengths;
  }
  if (!bestSize) return 0;
  return (
    bestSize +
    matchingCharacters(a, b, aLow, bestI, bLow, bestJ) +
    matchingCharacters(a, b, bestI + bestSize, aHigh, bestJ + bestSize, bHigh)
  );
}

function buildFeedback(
  total: number,
  lengthPoints: number,
  keywordPoints: number,
  conceptPoints: number,
  originalityPoints: number,
  coherencePoints: number
): string[] {
  const notes: string[] = [];
  if (total >= 70) {
    notes.push("Good recall. You captured enough of the passage to move ahead.");
  } else {
    notes.push("Try again and include the main idea plus two important details.");
  }
  if (lengthPoints < 15) notes.push("Write a little more so your answer has enough detail.");
  if (keywordPoints < 20) notes.push("Mention more important names, objects, or actions from the passage.");
  if (conceptPoints < 24) notes.push("Focus on what happened and what the passage teaches.");
  if (originalityPoints < 7) notes.push("Use your own words instead of copying the passage.");
  if (coherencePoints < 10) notes.push("Write complete, clear sentences.");
  return notes;
}
